### ISSUE TITLE
Limit broken with Group By in Scylla 5.4

### ISSUE THREAD

**Issue by tynli** [scylladb/scylladb#17237]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 5.4.2
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 20.04

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): physical
Hardware: sockets=1 cores=32 hyperthreading= memory=768GB
Disks: (SSD/HDD, count) NVMe, 9x, RAID-0

After upgrading from Scylla 5.2.14 to 5.4.2, `group by` queries with `limit` stopped limiting the results. I have been able to reproduce the issue with the following `cql`:

```
create table grouping (
  a int,
  b int,
  c text,
  PRIMARY KEY (a, b)
) WITH CLUSTERING ORDER BY (b ASC);
insert into grouping (a,b,c) values (1,2,'testing');
insert into grouping (a,b,c) values (1,3,'three');
insert into grouping (a,b,c) values (2,-1,'negative');
select * from grouping group by a limit 1;
```

Before the upgrade to 5.4.2, the snippet above would give the following results:

```
 a | b | c
---+---+---------
 1 | 2 | testing
```

After the upgrade, it gives the following:

```
 a | b  | c
---+----+----------
 1 |  2 |  testing
 2 | -1 | negative
```

I would expect the functionality of the `limit` statement to remain the same between versions (and have the `select` statement above produce a single output row).

**Comment by nyh** [comment]
We have two known issues about the combination of LIMIT and GROUP BY not working -
https://github.com/scylladb/scylladb/issues/5361
https://github.com/scylladb/scylladb/issues/5362

(one shows LIMIT returning too many rows, one shows too few rows!)

But both of these issues are old, so they wouldn't be a new regression in 5.4, so maybe this bug is slightly different.
We already have almost 20 different tests (!!) reproducing the aforementioned bugs, but I'll see if perhaps the issue seen here is slightly different and is worthy of a new test.

**Comment by nyh** [comment]
Unfortunately, I can confirm that the following test, in the context of `test/cql-pytest/test_group_by.py`, passes on Scylla 5.2.14 and on Cassandra, and fails on Scylla master. As the report says, the "LIMIT" seems to have no effect and two results are returned even though a "LIMIT 1" was used.
```python
# Adding a LIMIT should be honored.
# Reproduces #17237 - more results than the limit were generated.
def test_group_by_partition_with_limit(cql, table1):
    # "LIMIT 1" should only one of the two matches (0,0,0,1) and (1,0,0,5)
    # to be returned. The partition key 1 happens to be the first one in
    # murmur3 order, so this is what should be returned.
    assert {(1,0,0,5)} == set(cql.execute(f'SELECT p,c1,c2,v FROM {table1} GROUP BY p LIMIT 1'))
```

**Comment by nyh** [comment]
@mykaul I'll do a bisection to see what caused this regression.

**Comment by nyh** [comment]
Bisected the regression to commit 432cb02d643973aa0267a1dc42ccfe7feb99d153 by @avikivity:
```
432cb02d643973aa0267a1dc42ccfe7feb99d153 is the first bad commit
commit 432cb02d643973aa0267a1dc42ccfe7feb99d153
Author: Avi Kivity <avi@scylladb.com>
Date:   Fri Jun 23 21:55:50 2023 +0300

    cql3: select_statement: force aggregation if GROUP BY is used
    
    GROUP BY is typically used with aggregation. In one case the aggregation
    is implicit:
    
        SELECT a, b, c
        FROM tab
        GROUP BY x, y, z
    
    One row will appear from each group, even though no aggregation
    was specified. To avoid this irregularity, rewrite this query as
    
        SELECT first(a), first(b), first(c)
        FROM tab
        GROUP BY x, y, z
    
    This allows us to have different paths for aggregations and
    non-aggregations, without worrying about this special case.

 cql3/statements/select_statement.cc | 6 ++++++
 1 file changed, 6 insertions(+)
```

By the way, this is not the first GROUP BY regression that we've discovered in 5.4 caused by the same patch (part of the expression rewrite) - the previous one was https://github.com/scylladb/scylladb/issues/16531 fixed by https://github.com/scylladb/scylladb/pull/16559. Clearly our regression testing in this area did not cover enough cases :-( I'm constantly improved the CQL test suite, but wasn't good enough when the rewrite was done, and possibly still not good enough.

**Comment by nyh** [comment]
I suspect that since GROUP BY now uses "aggregation" (a silly function first(), but aggregation nontheless), we may be bitten the fact that LIMIT has a known bug with aggregation, where LIMIT is ignored - see https://github.com/scylladb/scylladb/issues/5361. Maybe it's finally time to fix that issue - that we know about for four years - and we'll get a fix for this issue as a byproduct.

**Comment by nyh** [comment]
@mykaul I think we're done with the triage - I found above a reproducing test, the source of the regression, and suspect that (since we don't want to revert Avi's expression rewrite), we need to fix #5361 to fix this issue.  Please let me know if you want me to go ahead and work on a fix, or you'd prefer to assign it to someone else.

**Comment by mykaul** [comment]
@kostja - please assign someone to look at this CQL regression.

**Comment by nyh** [comment]
> @kostja - please assign someone to look at this CQL regression.

I'll prepare a patch with the reproducing test that whoever works on this can use. Please also consider my suspicion that to fix this new regression we need to fix https://github.com/scylladb/scylladb/issues/5361, so the fix will hopefully also fix a bunch of existing tests mentioned there.

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/pull/17275 adds the cql-pytest test `test_group_by.py::test_group_by_partition_with_limit`.

**Comment by mykaul** [comment]
@avikivity will handle.

**Comment by mykaul** [comment]
> @avikivity will handle.

@avikivity - ping - was it handled?

**Comment by mykaul** [comment]
re-ping @avikivity - is this on track to 6.0?

**Comment by mykaul** [comment]
@avikivity - re-ping for 6.0?

**Comment by mykaul** [comment]
@avikivity / @ptrsmrn  - who should own this?

**Comment by mykaul** [comment]
> I think it should be me.

@KrzaQ - any updates? Is this still on 6.0?

**Comment by ptrsmrn** [comment]
> > I think it should be me.
> 
> @KrzaQ - any updates? Is this still on 6.0?

If we can wait another 1-3 weeks, then yes, otherwise should be rescheduled for 6.1.

**Comment by nyh** [comment]
> #17275 adds the cql-pytest test `test_group_by.py::test_group_by_partition_with_limit` that reproduces this regression. (Please also remember my comment above that #5361 mentions additional tests which might get fixed together with this one).

@KrzaQ when you are working on this issue (I see you have a draft #18842) please note this comment that I wrote a few months ago. I already wrote a very very simple (one line, using a table already created for another test) reproducer and it's already in the test suite (xfailing), but even more importantly - solving this issue should _hopefully_ fix several other xfailing tests (see #5361) and you should check that - and if it does report that it does in the commit message (and remove the "xfail" tag), and if it doesn't you may want to explain why and what remains to be fixed.

Thanks for working on this issue, it's pretty important in my own opinion - it's a correctness issue, and a regression, which is nasty.

**Comment by nyh** [comment]
> You're correct, a few of the xfail tests are passing now 🥳

Excellent - please see https://github.com/scylladb/scylladb/issues/5361#issuecomment-1459989637 for the full list. The tests which I mentioned as being written especially https://github.com/scylladb/scylladb/issues/5361 would hopefully start to pass (and their "xfail" line should be removed). A bunch of other tests which test several different things (unfortunately this is a common theme in tests translated from Cassandra's unit tests in test/cql-pytest/cassandra-tests) and exposed multiple bugs will continue to xfail, but you should nevertheless edit the xfail line to stop referring to this issue. 

Please also consider whether your patch fully fixes https://github.com/scylladb/scylladb/issues/5361 - and if so, add a "Fixes" referring to that issue as well. The reason why I predicted a solution would solve #5361 was that that issue was about the interaction of limit, groupby **and aggregation** - which in that old issue was count(), sum(), etc., but after 5.4, suddenly every group by became an "aggregation" (using a new "first()" aggregation) so this bug appeared also in cases that previously work.

And sadly, as I explained in the aforementioned comments, there is also the problem of dtests, which were "forced" into accepting the incorrect behavior of old versions of Scylla, and now either need to be fixed back to check for the correct behavior, or if we have the same test in cql-pytest/cassandra_tests, perhaps the dtests can even be removed (as I noted, about a decade ago Cassandra translated many of their old and slow dtests into quick CQL tests in Java, and I translated those into our Pytonn cql-pytest/cassandra_tests, so there is some overlap).

If any of the tests mentioned in https://github.com/scylladb/scylladb/issues/5361#issuecomment-1459989637 continue to fail after your patches, we should consider why, and perhaps close #5361 as fixed and open a new issue on the remaining bug.
